sc-memory/sc-memory/sc_type.cpp

```diff
@@ -6,65 +6,449 @@
+{
+  return sc_type_is_connector(m_realType);
+}
+
+bool ScType::IsCommonEdge() const
+{
+  return sc_type_is_common_edge(m_realType);
+}
+
+bool ScType::IsArc() const
+{
+  return sc_type_is_arc(m_realType);
+}
+
+bool ScType::IsCommonArc() const
+{
+  return sc_type_is_common_arc(m_realType);
+}
+
+bool ScType::IsMembershipArc() const
+{
+  return sc_type_is_membership_arc(m_realType);
+}
+
+bool ScType::IsNode() const
+{
+  return sc_type_is_node(m_realType);
+}
+
+bool ScType::IsLink() const
+{
+  return sc_type_is_node_link(m_realType);
+}
+
+bool ScType::IsConst() const
+{
+  return sc_type_has_subtype(m_realType, sc_type_const);
+}
+
+bool ScType::IsVar() const
+{
+  return sc_type_has_subtype(m_realType, sc_type_var);
+}
+
+bool ScType::IsUnknown() const
+{
+  return m_realType == 0;
+}
+
+bool ScType::HasConstancyFlag() const
+{
+  return sc_type_has_subtype_in_mask(m_realType, sc_type_constancy_mask);
+}
+
+ScType ScType::AsConst() const
+{
+  return ScType((m_realType & ~sc_type_var) | sc_type_const);
+}
+
+ScType ScType::UpConstType() const
+{
+  // TODO(NikitaZotov): implement meta-variables
+  return AsConst();
+}
+
+sc_type ScType::operator*() const
+{
+  return m_realType;
+}
+
+ScType & ScType::operator()(RealType bits)
+{
+  m_realType |= bits;
+  return *this;
+}
+
+bool ScType::operator==(ScType const & other)
+{
+  return m_realType == other.m_realType;
+}
+
+bool ScType::operator!=(ScType const & other)
+{
+  return m_realType != other.m_realType;
+}
+
+ScType::RealType ScType::BitAnd(RealType const & inMask) const
+{
+  return m_realType & inMask;
+}
+
+ScType ScType::operator|(ScType const & other)
+{
+  return ScType(m_realType | other.m_realType);
+}
+
+ScType ScType::operator&(ScType const & other)
+{
+  return ScType(m_realType & other.m_realType);
+}
+
+ScType & ScType::operator|=(ScType const & other)
+{
+  m_realType |= other.m_realType;
+  return *this;
+}
+
+ScType & ScType::operator&=(ScType const & other)
+{
+  m_realType &= other.m_realType;
+  return *this;
+}
+
+ScType::operator RealType() const
+{
+  return m_realType;
+}
+
+ScType::operator std::string() const
+{
+  auto const & it = m_typesToNames.find(*this);
+  if (it == m_typesToNames.cend())
+    return "";
+
+  return it->second;
+}
+
+std::string ScType::GetDirectSCsConnectorType() const
+{
+  return scs::TypeResolver::GetDirectSCsConnectorType(*this);
+}
+
+std::string ScType::GetReverseSCsConnectorType() const
+{
+  return scs::TypeResolver::GetReverseSCsConnectorType(*this);
+}
+
+std::string ScType::GetSCsElementKeynode() const
+{
+  return scs::TypeResolver::GetSCsElementKeynode(*this);
+}
+
+bool ScType::CanExtendTo(ScType const & extType) const
+{
+  return sc_memory_is_type_expendable_to(*this, extType);
+}
+
+std::set<ScType> ScType::GetConnectorTypes()
+{
+  std::set<ScType> connectorTypes;
+  for (auto const & [type, name] : m_typesToNames)
+  {
+    if (type.IsCommonEdge() || type.IsCommonArc() || type.IsMembershipArc())
```

Comment:
We don't allow creating a connector in SCs-code yet. because then we would have to implement the correct behaviour when specifying that a connector is an arc or an edge. The same for an element of unspecified type